gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,169 @@
+    ) -> pd.DataFrame:
+        """Calculate the total carbon generated based on the specified statistic
+        (min, median, max)"""
+        if stat not in ["min", "median", "max"]:
+            raise ValueError("stat must be one of 'min', 'median', or 'max'")
+
+        df = self._query_generation_per_plant(country_code, start, end)
+
+        generation_types = df.iloc[0].iloc[1:]
+        carbon_emission_per_plant = lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS[x][
+            stat
+        ]  # noqa: E731 C3001
+        emissions_map = generation_types.map(carbon_emission_per_plant)
+
+        df_numeric = df.iloc[2:].reset_index(drop=True)
+        df_numeric = df_numeric.set_index("Unnamed: 0").apply(  # noqa: E501
+            pd.to_numeric, errors="coerce"
+        )
+        df_numeric.index = pd.to_datetime(df_numeric.index)
+        df_numeric["Total Energy (kWh)"] = df_numeric.sum(axis=1) * 1000  # convert MWh to kWh
+        df_numeric["Total Carbon (gCO2eq)"] = df_numeric.mul(emissions_map, axis=1).sum(axis=1)
+
+        generation_types = df.iloc[0]
+        emissions_map = generation_types.map(
+            lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS.get(x, {}).get(stat, 0)
+        )
+        return df_numeric[["Total Energy (kWh)", "Total Carbon (gCO2eq)"]]
+
+    def update(self, country_code: str, time_slot: duration, results_list: List[Dict]):
+        """Update the results_dict with carbon generated and carbon savings"""
+
+        start_date = pd.Timestamp(results_list[0]["current_market"].split("T")[0])
+        end_date = pd.Timestamp(results_list[-1]["current_market"].split("T")[0])
+        df_total_carbon_and_energy = self._calculate_total_carbon_generated(
+            country_code, start_date, end_date
+        )
+        for result in results_list:
+            current_market = result["current_market"]
+
+            for area_uuid in result["results"].keys():
+                area_result = result["results"][area_uuid]
+                target_timestamp = pd.Timestamp(current_market).tz_localize("UTC")
+
+                nearest_index = df_total_carbon_and_energy.index.get_indexer(
+                    [target_timestamp], method="nearest"
+                )[0]
+                nearest_row = df_total_carbon_and_energy.iloc[nearest_index]
+
+                total_carbon_g = nearest_row["Total Carbon (gCO2eq)"]
+                total_energy_kWh = nearest_row["Total Energy (kWh)"]
+
+                imported_energy_from_grid_kWh = area_result["bought_from_grid"]
+                imported_energy_from_community_kWh = area_result[
+                    "energy_bought_from_community_kWh"
+                ]
+
+                carbon_generated_base_case = (
+                    total_carbon_g
+                    * (imported_energy_from_grid_kWh + imported_energy_from_community_kWh)
+                    / total_energy_kWh
+                )
+                carbon_generated_with_trading = (
```

Comment:
Minor, but we need to discuss the nomenclature here (talking about the suffix `_trading`). IMO we should be consistent use the suffix `_gsy` like in the savings calculation. Or, alternatively, `_p2p`, but this would also not be in line with the SCM.

Reply:
Using _gsy. Though in gsy we also perform non-p2p/non-trading scenarios